From 920162c1cff81c4ac5f9acb94de3d9559011072b Mon Sep 17 00:00:00 2001
From: Roopni Devanathan <quic_rdevanat@quicinc.com>
Date: Wed, 15 May 2024 09:35:48 +0530
Subject: [PATCH] wifi: ath12k: Fetch per-chain-RSSI from firmware

Currently, 'iw xxx station dump' command gives stats which
displays signal RSSI average. Along with the average signal RSSI,
per-chain RSSI values of each link should be printed. Firmware
gives these values. ath driver should pick up these values, parse
them over to be printed.

Dependency: [None]

Signed-off-by: Roopni Devanathan <quic_rdevanat@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h   |  3 +++
 drivers/net/wireless/ath/ath12k/dp_mon.c | 11 +++++++++++
 drivers/net/wireless/ath/ath12k/hal_rx.h | 13 +++++++++++--
 drivers/net/wireless/ath/ath12k/mac.c    | 24 +++++++++++++++++++++++-
 4 files changed, 48 insertions(+), 3 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -737,6 +737,7 @@ struct ath12k_link_sta {
 	u32 rx_retries;
 	u16 ast_hash;
 	u16 ast_idx;
+	s8 per_chain_rssi[IEEE80211_MAX_CHAINS];
 };
 
 #define ATH12K_REPORT_LOW_ACK_NUM_PKT_SSR 0xFFFF
@@ -801,6 +802,8 @@ enum ath12k_state {
 
 /* Antenna noise floor */
 #define ATH12K_DEFAULT_NOISE_FLOOR -95
+#define ATH12K_INVALID_RSSI_FULL -1
+#define ATH12K_INVALID_RSSI_EMPTY -128
 
 struct ath12k_ftm_event_obj {
 	u32 data_pos;
--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -1620,6 +1620,7 @@ ath12k_dp_mon_rx_parse_status_tlv(struct
 		break;
 
 	case HAL_PHYRX_RSSI_LEGACY: {
+		u8 i;
 		struct hal_rx_phyrx_rssi_legacy_info *rssi =
 			(struct hal_rx_phyrx_rssi_legacy_info *)tlv_data;
 		u32 rssi_legacy_info = __le32_to_cpu(rssi->info0);
@@ -1633,6 +1634,12 @@ ath12k_dp_mon_rx_parse_status_tlv(struct
 		ppdu_info->rssi_comb =
 			u32_get_bits(info[0],
 				     HAL_RX_PHYRX_RSSI_LEGACY_INFO_INFO1_RSSI_COMB);
+
+		for (i = 0; i < HAL_RX_MAX_NSS; i++)
+			ppdu_info->rssi_chain_pri20[i] =
+				le32_get_bits(rssi->preamble[i].rssi_2040,
+					      HAL_RX_PHYRX_RSSI_PREAMBLE_PRI20);
+
 		ppdu_info->bw = u32_get_bits(rssi_legacy_info,
 					     HAL_RX_PHYRX_RSSI_LEGACY_INFO_INFO0_RX_BW);
 		break;
@@ -3427,6 +3434,7 @@ static void ath12k_dp_mon_rx_update_peer
 	struct ath12k_rx_peer_stats *rx_stats = arsta->rx_stats;
 	u32 num_msdu;
 	u32 bw_offset;
+	u8 i;
 
 	arsta->last_tx_pkt_bw = ppdu_info->bw;
 	bw_offset = arsta->last_tx_pkt_bw * 3;
@@ -3482,6 +3490,9 @@ static void ath12k_dp_mon_rx_update_peer
 	rx_stats->num_mpdu_fcs_err += ppdu_info->num_mpdu_fcs_err;
 	rx_stats->dcm_count += ppdu_info->dcm;
 
+	for (i = 0; i < ARRAY_SIZE(arsta->per_chain_rssi); i++)
+		arsta->per_chain_rssi[i] = ppdu_info->rssi_chain_pri20[i];
+
 	rx_stats->rx_duration += ppdu_info->rx_duration;
 	arsta->rx_duration = rx_stats->rx_duration;
 
--- a/drivers/net/wireless/ath/ath12k/hal_rx.h
+++ b/drivers/net/wireless/ath/ath12k/hal_rx.h
@@ -224,6 +224,7 @@ struct hal_rx_mon_ppdu_info {
 	u8 ldpc;
 	u8 beamformed;
 	u8 rssi_comb;
+	u8 rssi_chain_pri20[HAL_RX_MAX_NSS];
 	u16 tid;
 	u8 fc_valid;
 	u16 ht_flags;
@@ -507,12 +508,20 @@ enum hal_rx_ul_reception_type {
 #define HAL_RX_PHYRX_RSSI_LEGACY_INFO_INFO0_RECEPTION   GENMASK(3, 0)
 #define HAL_RX_PHYRX_RSSI_LEGACY_INFO_INFO0_RX_BW	GENMASK(7, 5)
 #define HAL_RX_PHYRX_RSSI_LEGACY_INFO_INFO1_RSSI_COMB   GENMASK(15, 8)
+#define HAL_RX_PHYRX_RSSI_PREAMBLE_PRI20		GENMASK(7, 0)
+
+struct hal_rx_phyrx_chain_rssi {
+	__le32 rssi_2040;
+	__le32 rssi_80;
+} __packed;
 
 struct hal_rx_phyrx_rssi_legacy_info {
 	__le32 info0;
-	__le32 rsvd0[39];
+	__le32 rsvd0[7];
+	struct hal_rx_phyrx_chain_rssi pre_rssi[HAL_RX_MAX_NSS];
+	struct hal_rx_phyrx_chain_rssi preamble[HAL_RX_MAX_NSS];
 	__le32 info1;
-	__le32 rsvd2;
+	__le32 rsvd1;
 } __packed;
 
 #define HAL_RX_MPDU_START_INFO0_PPDU_ID		GENMASK(31, 16)
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -15588,6 +15588,11 @@ static void ath12k_mac_op_sta_statistics
 	struct ath12k_sta *ahsta = ath12k_sta_to_ahsta(sta);
 	struct ath12k_link_sta *arsta = &ahsta->deflink;
 	struct ath12k *ar = arsta->arvif->ar;
+	struct ath12k_base *ab = ar->ab;
+	bool db2dbm = test_bit(WMI_TLV_SERVICE_HW_DB2DBM_CONVERSION_SUPPORT,
+			       ab->wmi_ab.svc_map);
+	u8 i;
+	s8 rssi, rssi_offset, rssi_signal;
 
 	/* TODO accumulate link sta stats here? */
 
@@ -15618,7 +15623,26 @@ static void ath12k_mac_op_sta_statistics
 	sinfo->txrate.flags = arsta->txrate.flags;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_BITRATE);
 
-	sinfo->signal = arsta->rssi_comb + ar->rssi_offsets.rssi_offset;
+	for (i = 0; i < ARRAY_SIZE(sinfo->chain_signal); i++) {
+		sinfo->chains &= ~BIT(i);
+		rssi = arsta->per_chain_rssi[i];
+		ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
+			   "mac sta statistics rssi[%d] %d\n", i, rssi);
+		if (rssi != ATH12K_DEFAULT_NOISE_FLOOR &&
+		    rssi != ATH12K_INVALID_RSSI_FULL &&
+		    rssi != ATH12K_INVALID_RSSI_EMPTY &&
+		    rssi != 0) {
+			sinfo->chain_signal[i] = rssi;
+			sinfo->chains |= BIT(i);
+			sinfo->filled |= BIT_ULL(NL80211_STA_INFO_CHAIN_SIGNAL);
+		}
+	}
+
+	rssi_offset = arsta->rssi_comb + ar->rssi_offsets.rssi_offset;
+	rssi_signal = arsta->rssi_comb > ar->rssi_offsets.xlna_bypass_threshold ?
+		      rssi_offset + ar->rssi_offsets.xlna_bypass_offset :
+		      rssi_offset;
+	sinfo->signal = db2dbm ? arsta->rssi_comb : rssi_signal;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_SIGNAL);
 
 	sinfo->signal_avg = ewma_avg_rssi_read(&arsta->avg_rssi) +
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -13362,6 +13362,8 @@ ath12k_wmi_rssi_dbm_conv_subtlv_parser(s
 				min_nf = min(param_info->nf_hw_dbm[i][0], min_nf);
 		}
 		rssi_offsets->min_nf_dbm = min_nf;
+		rssi_offsets->xlna_bypass_offset = param_info->xlna_bypass_offset;
+		rssi_offsets->xlna_bypass_threshold = param_info->xlna_bypass_threshold;
 		break;
 	case WMI_TAG_RSSI_DBM_CONVERSION_TEMP_OFFSET_INFO:
 		if (len != sizeof(*temp_offset_info)) {
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -8558,6 +8558,8 @@ struct wmi_rssi_dbm_conv_offsets {
 	s8 min_nf_dbm;
 	/* rssi_offset is the sum of min_nf_dbm & rssi_temp_offset*/
 	s32 rssi_offset;
+	u32 xlna_bypass_offset;
+	u32 xlna_bypass_threshold;
 };
 
 #define WMI_LATENCY_TID_INFO_SCS_TID_NUM		GENMASK(7, 0)
